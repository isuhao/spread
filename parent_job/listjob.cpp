#include "listjob.hpp"
#include <assert.h>
#include <job/thread.hpp>

#ifdef NEED_LOCKGUARD
#include <boost/thread/lock_guard.hpp>
#endif

#define LOCK boost::lock_guard<boost::mutex> lock(mutex)

using namespace Spread;

void ListJob::add(JobPtr j)
{
  LOCK;
  jobs.insert(j);
  assert(j);
  assert(!j->getInfo()->hasStarted());
  if(started) Thread::run(j);
}

void ListJob::start()
{
  assert(info->isBusy());
  if(started) return;

  LOCK;

  // Tell add() to start newly added jobs immediately
  started = true;

  // Start all jobs already in the list
  JobSet::const_iterator it;
  for(it = jobs.begin(); it != jobs.end(); it++)
    {
      const JobPtr &j = *it;
      assert(j);
      assert(!j->getInfo()->hasStarted());
      Thread::run(j);
    }
}
