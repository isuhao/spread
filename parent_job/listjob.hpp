#ifndef __SPREAD_LIST_JOB_HPP_
#define __SPREAD_LIST_JOB_HPP_

#include "parentjob.hpp"

/* ListJob is a ParentJob that has multiple child jobs, each running
   in a separate thread.

   New jobs can be added with add() before or after the job has
   started. The list is thread safe, meaning you can safely add new
   jobs from external threads.

   This is an abstract class. For usable implementations see AndJob
   and JobHolder.
 */

namespace Spread
{
  struct ListJob : ParentJob
  {
    /* Add a new child job to the list. The job must be freshly
       created (not started yet). The job is started by start(), or
       immediately if start() has already been executed.
     */
    void add(JobPtr j);

    // Convenience pointer version. Only use on pointers NOT owned by
    // a shared_ptr / JobPtr.
    void add(Job *j) { add(JobPtr(j)); }

    ListJob() : started(false) {}

  protected:
    /* Call start() from doJob() whenever you are ready to start the
       jobs. This will start all jobs already added to the list, and
       will also make add() auto-start any jobs added from this point
       on.
     */
    void start();

  private:
    bool started;
  };
}
#endif
