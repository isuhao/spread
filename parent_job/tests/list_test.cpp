#include "listjob.hpp"
#include "common.cpp"

struct ListTest : ListJob
{
  int i;

  ListTest(int _i) : i(_i) {}

  void doJob()
  {
    cout << "Starting list " << i << endl;
    start();
    Thread::sleep(0.5);
    setDone();
  }
};

int main()
{
  ListJob *lst = new ListTest(0);
  JobPtr ptr(lst);

  cout << "\nADDING BEFORE START:\n";
  lst->add(JobPtr(new ListTest(1)));
  lst->add(new ListTest(2));
  print(ptr);

  cout << "\nSTARTING NOW:\n";
  ptr->run();
  print(ptr);

  cout << "\nADDING MORE STUFF:\n";
  lst->add(JobPtr(new ListTest(3)));
  lst->add(new ListTest(4));
  Thread::sleep(1.0);
  print(ptr);

  return 0;
}
