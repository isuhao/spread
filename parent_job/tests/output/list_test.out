
ADDING BEFORE START:
UNK:  0/0
  JOBS:
  UNK:  0/0
  UNK:  0/0

STARTING NOW:
Starting list 0
Starting list 1
Starting list 2
SUC: Success 0/0
  JOBS:
  BUS:  0/0
  BUS:  0/0

ADDING MORE STUFF:
Starting list 4
Starting list 3
SUC: Success 0/0
  JOBS:
  SUC: Success 0/0
  SUC: Success 0/0
  SUC: Success 0/0
  SUC: Success 0/0
